File: dsa/li13_graph/MultiBfs_L994_RottingOranges.java
package li13_graph;

import java.util.LinkedList;
import java.util.Queue;

public class MultiBfs_L994_RottingOranges {

	public static void main(String[] args) {
		MultiBfs_L994_RottingOranges obj = new MultiBfs_L994_RottingOranges();
//		int[][] grid = new int[][]{
//			{2,1,1},
//			{1,1,0},
//			{0,1,1}};
		int[][] grid = new int[][]{
			{1,2}};
		System.out.println(obj.orangesRotting(grid));
	}

	// Problem: https://leetcode.com/problems/rotting-oranges/
    // 1 -> fresh, 2 -> rotten, 0 -> empty
    // Idea: We have to multi source bfs and increase time for every level.
    // We need to count the fresh nodes and load all the roatten oranges to queue
    // In every step, for every orange in queue, check in its 4 direction, if you find a rotten orange make it rotten, reduce the fresh counter and load the rotten one to queue.
    // Run this algo until there no more fresh oranges, or there no rotten orange can reach remaining fresh one's.
	// Time: O(n*m), Space: O(n*m)
    public int orangesRotting(int[][] grid) {
        int m = grid.length, n = grid[0].length;

        int[][] vis = new int[m][n];
        int fresh = 0;
        Queue<int[]> q = new LinkedList<>();
        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                if(grid[i][j] == 1)
                    fresh++;
                if(grid[i][j] == 2){
                    q.add(new int[] {i,j,0});   // x,y,mins
                    vis[i][j] = 2;              // we will use 2 for visited/rottened
                }
            }
        }

        int maxMins = 0;
        int[][] dirs = new int[][]{{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        while(!q.isEmpty()){
            int[] curr = q.remove();
            int row = curr[0], col = curr[1], mins = curr[2];
            maxMins = Math.max(maxMins, mins); // max of all minutes

            for (int[] dir : dirs) {
                int nrow = row + dir[0];
                int ncol = col + dir[1];
                if (isSafe(grid, nrow, ncol, vis)) {
                    vis[nrow][ncol] = 2;
                    q.add(new int[] { nrow, ncol, mins + 1 });  // pushing mins to q so that we don't need to do level wise traversal
                    fresh--;
                }
            }
        }
        return (fresh == 0)? maxMins: -1;
    }

    private boolean isSafe(int[][] grid, int i, int j, int[][] vis){
        return i>=0 && i<grid.length && j>=0 && j<grid[0].length && grid[i][j] == 1 && vis[i][j] != 2;
    }
}

File: dsa/patterns/13_graphs/python/multi_bfs_l994_rotting_oranges.py
from collections import deque


def oranges_rotting(grid):
    """
    Problem: https://leetcode.com/problems/rotting-oranges/

    1 -> fresh, 2 -> rotten, 0 -> empty

    Idea:
    - This is a **multi-source BFS** problem.
    - Count all fresh oranges initially.
    - Push all rotten oranges into the queue with time = 0.
    - For each rotten orange, try to rot its 4-direction neighbors.
    - Each newly rotten orange is added to the queue with time + 1.
    - Track the maximum time taken.
    - If fresh oranges remain at the end, return -1.

    Time: O(m * n)
    Space: O(m * n)
    """
    m, n = len(grid), len(grid[0])

    fresh = 0
    visited = [[False] * n for _ in range(m)]
    q = deque()

    # Initialize queue with all rotten oranges
    for i in range(m):
        for j in range(n):
            if grid[i][j] == 1:
                fresh += 1
            elif grid[i][j] == 2:
                q.append((i, j, 0))   # (row, col, minutes)
                visited[i][j] = True

    directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
    max_minutes = 0

    while q:
        row, col, minutes = q.popleft()
        max_minutes = max(max_minutes, minutes)

        for dr, dc in directions:
            nr, nc = row + dr, col + dc
            if (
                0 <= nr < m and
                0 <= nc < n and
                grid[nr][nc] == 1 and
                not visited[nr][nc]
            ):
                visited[nr][nc] = True
                fresh -= 1
                q.append((nr, nc, minutes + 1))

    return max_minutes if fresh == 0 else -1


if __name__ == "__main__":
    grid = [
        [1, 2]
    ]
    print(oranges_rotting(grid))
    